sway/tree/container.c

```diff
@@ -106,103 +106,209 @@ static struct sway_container *_container_destroy(struct sway_container *cont) {
-	struct output_config *oc = NULL, *all = NULL;
-	for (int i = 0; i < config->output_configs->length; ++i) {
-		struct output_config *cur = config->output_configs->items[i];
+	wl_list_remove(&output->sway_output->damage_destroy.link);
+	wl_list_remove(&output->sway_output->damage_frame.link);
 
-		if (strcasecmp(name, cur->name) == 0 ||
-				strcasecmp(identifier, cur->name) == 0) {
-			wlr_log(L_DEBUG, "Matched output config for %s", name);
-			oc = cur;
-		}
-		if (strcasecmp("*", cur->name) == 0) {
-			wlr_log(L_DEBUG, "Matched wildcard output config for %s", name);
-			all = cur;
+	wlr_log(L_DEBUG, "OUTPUT: Destroying output '%s'", output->name);
+	_container_destroy(output);
+	return &root_container;
+}
+
+static struct sway_container *container_workspace_destroy(
+		struct sway_container *workspace) {
+	if (!sway_assert(workspace, "cannot destroy null workspace")) {
+		return NULL;
+	}
+
+	// Do not destroy this if it's the last workspace on this output
+	struct sway_container *output = container_parent(workspace, C_OUTPUT);
+	if (output && output->children->length == 1) {
+		return NULL;
+	}
+
+	struct sway_container *parent = workspace->parent;
+	if (workspace->children->length == 0) {
+		// destroy the WS if there are no children (TODO check for floating)
+		wlr_log(L_DEBUG, "destroying workspace '%s'", workspace->name);
+		ipc_event_workspace(workspace, NULL, "empty");
+	} else {
+		// Move children to a different workspace on this output
+		struct sway_container *new_workspace = NULL;
+		// TODO move floating
+		for (int i = 0; i < output->children->length; i++) {
+			if (output->children->items[i] != workspace) {
+				new_workspace = output->children->items[i];
+				break;
+			}
 		}
 
-		if (oc && all) {
-			break;
+		wlr_log(L_DEBUG, "moving children to different workspace '%s' -> '%s'",
+			workspace->name, new_workspace->name);
+		for (int i = 0; i < workspace->children->length; i++) {
+			container_move_to(workspace->children->items[i], new_workspace);
 		}
 	}
-	if (!oc) {
-		oc = all;
+
+	_container_destroy(workspace);
+
+	output_damage_whole(output->sway_output);
+
+	return parent;
+}
+
+static void container_root_finish(struct sway_container *con) {
+	wlr_log(L_ERROR, "TODO: destroy the root container");
+}
+
+static bool container_reap_empty(struct sway_container *con) {
+	switch (con->type) {
+	case C_ROOT:
+	case C_OUTPUT:
+		// dont reap these
+		break;
+	case C_WORKSPACE:
+		if (!workspace_is_visible(con) && con->children->length == 0) {
+			container_workspace_destroy(con);
+			return true;
+		}
+		break;
+	case C_CONTAINER:
+		if (con->children->length == 0) {
+			_container_destroy(con);
+			return true;
+		} else if (con->children->length == 1) {
+			struct sway_container *child = con->children->items[0];
+			if (child->type == C_CONTAINER) {
+				container_remove_child(child);
+				container_replace_child(con, child);
+				_container_destroy(con);
+				return true;
+			}
+		}
+	case C_VIEW:
+		break;
+	case C_TYPES:
```

Comment:
We wanted to get rid of this `C_TYPES` right?

Reply:
Yeah, we can do that in another PR. This is only for closing.